File: src/lib/preventiveIntelligence.ts
import "server-only";
import { prisma } from "@/lib/prisma";
import { computeTrendEngine, type TrendDataPoint, type TrendIndicator } from "@/lib/trendEngine";
import { computeSobrecargaProbability, computeSubutilizacionPredictions, computeProjectDelayPrediction, computeOperationalStability } from "@/lib/predictionEngine";

/**
 * Inteligencia Preventiva (Sprint E, Bloque 7) — compone las salidas YA
 * CALCULADAS de trendEngine.ts/predictionEngine.ts en una lista priorizada
 * de alertas legibles. Explícitamente SEPARADA de `computeAlerts`
 * (analytics.ts, motor de 8 reglas) y de `riskAlerts.ts` (vestigial) —
 * ninguno de los dos se importa para mutar ni se toca.
 */
export type PreventiveSeverity = "roja" | "naranja" | "amarilla" | "verde";

export type PreventiveAlert = {
  severity: PreventiveSeverity;
  message: string;
  source: string;
  relatedIndicator: TrendIndicator | "sobrecarga" | "subutilizacion" | "retraso" | "estabilidad";
};

const SEVERITY_RANK: Record<PreventiveSeverity, number> = { roja: 4, naranja: 3, amarilla: 2, verde: 1 };

function sortBySeverity(alerts: PreventiveAlert[]): PreventiveAlert[] {
  return alerts.slice().sort((a, b) => SEVERITY_RANK[b.severity] - SEVERITY_RANK[a.severity]);
}

/** Semanas consecutivas (desde la más reciente hacia atrás) en que el valor bajó respecto a la anterior. */
function consecutiveDeclineWeeks(points: TrendDataPoint[]): number {
  let streak = 0;
  for (let i = points.length - 1; i > 0; i--) {
    if (points[i].value < points[i - 1].value) streak++;
    else break;
  }
  return streak > 0 ? streak + 1 : 0; // +1: el streak cuenta caídas, la "semana" incluye el punto de partida.
}

const ORDINAL: Record<number, string> = { 1: "primera", 2: "segunda", 3: "tercera", 4: "cuarta", 5: "quinta" };

export async function computePreventiveAlerts(userId: string, now: Date = new Date()): Promise<PreventiveAlert[]> {
  const [sobrecarga, trend, stability, subutilizacionMap] = await Promise.all([
    computeSobrecargaProbability(userId, now),
    computeTrendEngine(userId, now),
    computeOperationalStability(userId, now),
    computeSubutilizacionPredictions([userId], now),
  ]);

  const alerts: PreventiveAlert[] = [];

  if (sobrecarga.available && sobrecarga.nivel !== "Bajo") {
    alerts.push({
      severity: sobrecarga.nivel === "Alto" ? "roja" : "naranja",
      message:
        sobrecarga.horizon <= 7
          ? `Existe ${sobrecarga.nivel === "Alto" ? "alta" : "media"} probabilidad de sobrecarga la próxima semana.`
          : `Existe ${sobrecarga.nivel === "Alto" ? "alta" : "media"} probabilidad de sobrecarga en los próximos ${sobrecarga.horizon} días.`,
      source: "Predicción de Sobrecarga",
      relatedIndicator: "sobrecarga",
    });
  }

  const cumplimiento = trend.indicators.cumplimiento;
  if (cumplimiento.available && cumplimiento.direction === "negativa") {
    const streak = consecutiveDeclineWeeks(cumplimiento.dataPoints);
    if (streak >= 2) {
      alerts.push({
        severity: streak >= 3 ? "roja" : "naranja",
        message: `La tendencia de cumplimiento disminuye por ${ORDINAL[Math.min(streak, 5)] ?? `${streak}ª`} semana consecutiva.`,
        source: "Trend Engine — Cumplimiento",
        relatedIndicator: "cumplimiento",
      });
    }
  }

  const subutilizacion = subutilizacionMap.get(userId);
  if (subutilizacion && subutilizacion.nivel === "Alto") {
    alerts.push({
      severity: "amarilla",
      message: "Presenta una proyección de subutilización de capacidad.",
      source: "Predicción de Subutilización",
      relatedIndicator: "subutilizacion",
    });
  }

  if (stability.classification === "Baja" || stability.classification === "Muy Baja") {
    alerts.push({
      severity: stability.classification === "Muy Baja" ? "naranja" : "amarilla",
      message: `Estabilidad operativa ${stability.classification.toLowerCase()} — variabilidad significativa en: ${stability.basedOn.join(", ") || "varios indicadores"}.`,
      source: "Estabilidad Operativa",
      relatedIndicator: "estabilidad",
    });
  }

  if (alerts.length === 0) {
    return [{ severity: "verde", message: "Sin riesgos preventivos detectados.", source: "Inteligencia Preventiva", relatedIndicator: "estabilidad" }];
  }
  return sortBySeverity(alerts);
}

/**
 * Vista de equipo — mismos bloques de predicción, en lote (batch de
 * `computeSubutilizacionPredictions` ya evita N+1; `computeProjectDelayPrediction`
 * se llama una vez por proyecto visible, no por participante).
 */
export async function computeTeamPreventiveAlerts(userIds: string[], projectIds: string[], now: Date = new Date()): Promise<PreventiveAlert[]> {
  const [subutilizacionMap, users, projectDelays, projects] = await Promise.all([
    computeSubutilizacionPredictions(userIds, now),
    prisma.user.findMany({ where: { id: { in: userIds } }, select: { id: true, name: true } }),
    Promise.all(projectIds.map((id) => computeProjectDelayPrediction(id, now))),
    prisma.project.findMany({ where: { id: { in: projectIds } }, select: { id: true, name: true } }),
  ]);

  const userName = new Map(users.map((u) => [u.id, u.name]));
  const projectName = new Map(projects.map((p) => [p.id, p.name]));
  const alerts: PreventiveAlert[] = [];

  for (const [userId, prediction] of subutilizacionMap) {
    if (prediction.nivel === "Alto") {
      alerts.push({
        severity: "amarilla",
        message: `${userName.get(userId) ?? "Un colaborador"} presenta una proyección de subutilización.`,
        source: "Predicción de Subutilización",
        relatedIndicator: "subutilizacion",
      });
    }
  }

  projectIds.forEach((projectId, i) => {
    const prediction = projectDelays[i];
    if (prediction.available && prediction.nivel !== "Bajo") {
      alerts.push({
        severity: prediction.nivel === "Alto" ? "roja" : "naranja",
        message: `Existe riesgo ${prediction.nivel === "Alto" ? "elevado" : "moderado"} de retraso en ${projectName.get(projectId) ?? "un proyecto"}.`,
        source: "Predicción de Retrasos",
        relatedIndicator: "retraso",
      });
    }
  });

  if (alerts.length === 0) {
    return [{ severity: "verde", message: "Sin riesgos preventivos detectados en el equipo.", source: "Inteligencia Preventiva", relatedIndicator: "estabilidad" }];
  }
  return sortBySeverity(alerts);
}

File: backend/apps/analytics/preventive_intelligence.py
"""Inteligencia Preventiva — Fase 9b (ver docs/AUDIT_LOG.md §
2026-08-18). Réplica exacta de `src/lib/preventiveIntelligence.ts`:
compone las salidas YA CALCULADAS de `trend_engine.py`/
`prediction_engine.py` en una lista priorizada de alertas legibles.
Explícitamente SEPARADA del motor de 8 reglas de `pipeline.py`/
`alerts_engine.py` — ninguno de los dos se importa para mutar ni se
toca."""

from datetime import datetime
from datetime import timezone as dt_timezone

from apps.projects.models import Project
from apps.users.models import User

from .prediction_engine import (
    compute_operational_stability,
    compute_project_delay_prediction,
    compute_sobrecarga_probability,
    compute_subutilizacion_predictions,
)
from .trend_engine import compute_trend_engine

SEVERITY_RANK = {"roja": 4, "naranja": 3, "amarilla": 2, "verde": 1}

ORDINAL = {1: "primera", 2: "segunda", 3: "tercera", 4: "cuarta", 5: "quinta"}


def _sort_by_severity(alerts: list[dict]) -> list[dict]:
    return sorted(alerts, key=lambda a: SEVERITY_RANK[a["severity"]], reverse=True)


def _consecutive_decline_weeks(points: list[dict]) -> int:
    """Semanas consecutivas (desde la más reciente hacia atrás) en que
    el valor bajó respecto a la anterior. Réplica exacta de
    `consecutiveDeclineWeeks`."""
    streak = 0
    for i in range(len(points) - 1, 0, -1):
        if points[i]["value"] < points[i - 1]["value"]:
            streak += 1
        else:
            break
    return streak + 1 if streak > 0 else 0  # +1: el streak cuenta caídas, la "semana" incluye el punto de partida.


def compute_preventive_alerts(*, user, now: datetime | None = None) -> list[dict]:
    """Réplica exacta de `computePreventiveAlerts`."""
    now = now or datetime.now(dt_timezone.utc)
    sobrecarga = compute_sobrecarga_probability(user=user, now=now)
    trend = compute_trend_engine(user=user, now=now)
    stability = compute_operational_stability(user=user, now=now)
    subutilizacion_map = compute_subutilizacion_predictions(user_ids=[user.id], now=now)

    alerts: list[dict] = []

    if sobrecarga["available"] and sobrecarga["nivel"] != "Bajo":
        alto = sobrecarga["nivel"] == "Alto"
        alerts.append(
            {
                "severity": "roja" if alto else "naranja",
                "message": (
                    f"Existe {'alta' if alto else 'media'} probabilidad de sobrecarga la próxima semana."
                    if sobrecarga["horizon"] <= 7
                    else f"Existe {'alta' if alto else 'media'} probabilidad de sobrecarga en los próximos {sobrecarga['horizon']} días."
                ),
                "source": "Predicción de Sobrecarga",
                "related_indicator": "sobrecarga",
            }
        )

    cumplimiento = trend["indicators"]["cumplimiento"]
    if cumplimiento["available"] and cumplimiento["direction"] == "negativa":
        streak = _consecutive_decline_weeks(cumplimiento["data_points"])
        if streak >= 2:
            ordinal = ORDINAL.get(min(streak, 5), f"{streak}ª")
            alerts.append(
                {
                    "severity": "roja" if streak >= 3 else "naranja",
                    "message": f"La tendencia de cumplimiento disminuye por {ordinal} semana consecutiva.",
                    "source": "Trend Engine — Cumplimiento",
                    "related_indicator": "cumplimiento",
                }
            )

    subutilizacion = subutilizacion_map.get(user.id)
    if subutilizacion and subutilizacion["nivel"] == "Alto":
        alerts.append(
            {
                "severity": "amarilla",
                "message": "Presenta una proyección de subutilización de capacidad.",
                "source": "Predicción de Subutilización",
                "related_indicator": "subutilizacion",
            }
        )

    if stability["classification"] in ("Baja", "Muy Baja"):
        alerts.append(
            {
                "severity": "naranja" if stability["classification"] == "Muy Baja" else "amarilla",
                "message": (
                    f"Estabilidad operativa {stability['classification'].lower()} — variabilidad significativa en: "
                    f"{', '.join(stability['based_on']) or 'varios indicadores'}."
                ),
                "source": "Estabilidad Operativa",
                "related_indicator": "estabilidad",
            }
        )

    if not alerts:
        return [
            {
                "severity": "verde",
                "message": "Sin riesgos preventivos detectados.",
                "source": "Inteligencia Preventiva",
                "related_indicator": "estabilidad",
            }
        ]
    return _sort_by_severity(alerts)


def compute_team_preventive_alerts(*, user_ids: list[int], project_ids: list[int], now: datetime | None = None) -> list[dict]:
    """Vista de equipo — mismos bloques de predicción, en lote (batch
    de `compute_subutilizacion_predictions` ya evita N+1;
    `compute_project_delay_prediction` se llama una vez por proyecto
    visible, no por participante). Réplica exacta de
    `computeTeamPreventiveAlerts`."""
    now = now or datetime.now(dt_timezone.utc)
    subutilizacion_map = compute_subutilizacion_predictions(user_ids=user_ids, now=now)
    user_names = {u.id: u.first_name for u in User.objects.filter(id__in=user_ids).only("id", "first_name")}
    project_delays = [compute_project_delay_prediction(project_id=pid, now=now) for pid in project_ids]
    project_names = {p.id: p.name for p in Project.objects.filter(id__in=project_ids).only("id", "name")}

    alerts: list[dict] = []

    for user_id, prediction in subutilizacion_map.items():
        if prediction["nivel"] == "Alto":
            name = user_names.get(user_id)
            alerts.append(
                {
                    "severity": "amarilla",
                    "message": f"{name if name is not None else 'Un colaborador'} presenta una proyección de subutilización.",
                    "source": "Predicción de Subutilización",
                    "related_indicator": "subutilizacion",
                }
            )

    for project_id, prediction in zip(project_ids, project_delays, strict=False):
        if prediction.get("available") and prediction["nivel"] != "Bajo":
            name = project_names.get(project_id)
            alerts.append(
                {
                    "severity": "roja" if prediction["nivel"] == "Alto" else "naranja",
                    "message": (
                        f"Existe riesgo {'elevado' if prediction['nivel'] == 'Alto' else 'moderado'} de retraso en "
                        f"{name if name is not None else 'un proyecto'}."
                    ),
                    "source": "Predicción de Retrasos",
                    "related_indicator": "retraso",
                }
            )

    if not alerts:
        return [
            {
                "severity": "verde",
                "message": "Sin riesgos preventivos detectados en el equipo.",
                "source": "Inteligencia Preventiva",
                "related_indicator": "estabilidad",
            }
        ]
    return _sort_by_severity(alerts)
